sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -549,6 +549,96 @@ case class CumeDist() extends RowNumberLike with SizeBasedWindowFunction {
   override def prettyName: String = "cume_dist"
 }
 
+/**
+ * The NthValue function returns the value of `input` at the `offset`th row from beginning of the
+ * window frame. Offset starts at 1. When the value of `input` is null at the `offset`th row or
+ * there is no such an `offset`th row, null is returned.
+ */
+@ExpressionDescription(
+  usage = """
+    _FUNC_(input[, offset]) - Returns the value of `input` at the row that is the `offset`th row
+      from beginning of the window frame. Offsets start at 1. If the value of `input` at the
+      `offset`th row is null, null is returned. If there is no such an offset row (e.g., when the
+      offset is 10, size of the window frame less than 10), null is returned.
+  """,
+  arguments = """
+    Arguments:
+      * input - the target column or expression that the function operates on.
+      * offset - an int expression which determines the row number relative to the first row in
+          the window for which to return the expression. The offset can be a constant or an
+          expression and must be a positive integer that is greater than 0.
+      * ignoreNulls - an optional specification that indicates the NthValue should skip null
+          values in the determination of which row to use.
+  """,
+  since = "3.1.0",
+  group = "window_funcs")
+case class NthValue(input: Expression, offsetExpr: Expression, ignoreNulls: Boolean)
```

Comment:
can we add a TODO to optimize it using `OffsetWindowFunction`?

Reply:
BTW, which window frames does `OffsetWindowFunction` support?

Reply:
`OffsetWindowFunction` not support any specified frame now.
`OffsetWindowFunction`  limit it with
```
  override lazy val frame: WindowFrame = {
    val boundary = direction match {
      case Ascending => offset
      case Descending => UnaryMinus(offset) match {
          case e: Expression if e.foldable => Literal.create(e.eval(EmptyRow), e.dataType)
          case o => o
      }
    }
    SpecifiedWindowFrame(RowFrame, boundary, boundary)
  }
```